### PR TITLE
Avatar implementation

### PR DESCRIPTION
**Pull request by YeahImRose**
A squashed version of PR #92

### CODE REVIEW

File: osu.Game/Graphics/Avatar.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+using osu.Framework.Graphics.OpenGL;
+using System.Net;
+using System.Collections.Generic;
+
+namespace osu.Game.Graphics
+{
+    public class Avatar : Sprite
+    {
+        private int userId;
+        private int avatarSize = 50;
```
Comment (peppy) [inline osu.Game/Graphics/Avatar.cs]: this is silly. why not just use .Size or .Scale?

File: osu.Game/Overlays/Toolbar.cs [hunk 2]
```diff
@@ -87,26 +121,36 @@ public override void Load(BaseGame game)
...
                         {
                             Icon = FontAwesome.bars
                         },
                     }
                 }
             };
+
+            InitUser();
+            Scheduler.AddDelayed(delegate {
+                InitUser();
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar.cs]: what?

File: osu.Game/Overlays/Toolbar.cs [hunk 3]
```diff
@@ -39,13 +47,39 @@ protected override void PopOut()
...
+            else if (user.UserId != toolbarUserButton.UserId)
+            {
+                toolbarUserButton.UpdateButton(user.UserId, user.Name);
+                toolbarUserButton.UserId = user.UserId;
+            }
+        }
 
-            RelativeSizeAxes = Axes.X;
-            Size = new Vector2(1, height);
+        private void InitUser()
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar.cs]: you're calling this every 10 seconds with no checks for whether it's still running, failure of last call, etc.

what?
what?
what?

File: osu.Game/Overlays/Toolbar.cs [hunk 4]
```diff
@@ -39,13 +47,39 @@ protected override void PopOut()
...
+        {
+            MyUser req = new MyUser();
+            api.Queue(req);
+            req.Success += delegate (User MyUser)
+            {
+                this.user = MyUser;
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar.cs]: you're throwing out the old user and replacing it every time this is called. please no.

File: osu.Game/Graphics/Avatar.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            string url = "https://a.ppy.sh/" + userid.ToString();
+
+            byte[] imageData = null;
+
+            if (userid != 1)
+            {
+                try
+                {
+                    using (var wc = new System.Net.WebClient())
```
Comment (peppy) [inline osu.Game/Graphics/Avatar.cs]: why are you blatantly ignoring the code i wrote for async retrieval, which also has error handling, timeout checking and protocol fallback support?

also there's no debounce

File: osu.Game/Graphics/Avatar.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                        imageData = await wc.DownloadDataTaskAsync(new System.Uri(url));
+                }
+                catch { }
+            }
+            Scheduler.Add(delegate
+            {
+                if (imageData != null)
+                {
+                    Texture = TextureLoader.FromBytes(imageData);
```
Comment (peppy) [inline osu.Game/Graphics/Avatar.cs]: not using atlases

File: osu.Game/Graphics/Avatar.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                }
+            });
+        }
+
+        public override void Load(BaseGame game)
+        {
+            this.game = game;
```
Comment (peppy) [inline osu.Game/Graphics/Avatar.cs]: why? you never use it?

File: osu.Game/Overlays/Toolbar.cs [hunk 8]
```diff
@@ -87,26 +121,36 @@ public override void Load(BaseGame game)
...
+            }, 10000, true);
+
+            RelativeSizeAxes = Axes.X;
+            Size = new Vector2(1, height);
         }
 
+        protected override void Update()
+        {
+            base.Update();
+            api.Update();
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar.cs]: this is not intended use. you're taking Game's api and then updating it here a second time.

File: osu.Game/Overlays/Toolbar.cs [hunk 9]
```diff
@@ -87,26 +121,36 @@ public override void Load(BaseGame game)
...
                             Icon = FontAwesome.bars
                         },
                     }
                 }
             };
+
+            InitUser();
+            Scheduler.AddDelayed(delegate {
+                InitUser();
+            }, 10000, true);
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar.cs]: scheduling with no cancel logic?

File: osu.Game/Overlays/Toolbar.cs [hunk 10]
```diff
@@ -39,13 +47,39 @@ protected override void PopOut()
...
+                CheckUser();
+            };
+        }
 
+
+        public override void Load(BaseGame game)
+        {
+            base.Load(game);
+            api = ((OsuGameBase)game).API;
+            this.game = game;
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar.cs]: not using this, don't store.

**Comment by peppy** [comment]
you don't need to add "squashed" in commit messages.
also you should work on explaining what you are doing in the PR better.
also you know you can rebase/force push rather than opening a new PR, right?

**Comment by Lunar-Dust** [comment]
I had over 130 commits that I'd have to manually add due to merge conflicts. I know rebasing is a thing. That's what I tried, and it was literally going to take hours. This was a much faster solution. I'll do some more work on this and update it later this week probably.

**Comment by peppy** [comment]
i meant more that you could force push your new head on your old PR, not that you interactive-rebase one step at a time.

**Comment by peppy** [comment]
please fix your commits and come back if/when done.

**Comment by Lunar-Dust** [comment]
Re-did my commits and made intelligent commit messages, down to 6 commits now

**Pull request closed without merging**